contexts/DelegationContext.tsx

```diff
@@ -0,0 +1,292 @@
+    );
+  }
+
+  function getDelegateAddress() {
+    const status = getDelegationStatus();
+    if (status === "delegator") return delegate;
+    if (status === "delegate") return address;
+    return zeroAddress;
+  }
+
+  function getDelegatorAddress() {
+    const status = getDelegationStatus();
+    if (status === "delegator") return address;
+    if (status === "delegate") return voterFromDelegate;
+    return zeroAddress;
+  }
+
+  function getDelegationStatus(): DelegationStatusT {
+    if (!address) return "no-wallet-connected";
+    // if you have neither `DelegatorSet` nor `DelegateSet` events, you are neither a delegate or a delegator
+    if (
+      !getHasReceivedRequestsToBeDelegate() &&
+      !getHasPendingSentRequestsToBeDelegate() &&
+      !getHasDelegatorSetEvents()
+    )
+      return "no-delegation";
+    // if there is a delegator set for your address, you are a delegate
+    if (voterFromDelegate.toLowerCase() !== address.toLowerCase()) return "delegate";
+    // if the `delegateToStaker` mapping for the `delegate` defined in your `voterStakes`, then you are a delegator
+    if (address.toLowerCase() === delegateToStaker.toLowerCase()) return "delegator";
+    // if the user has received a request to be another wallet's delegate but they have not accepted any, then they are a pending delegate
+    if (getHasPendingReceivedRequestsToBeDelegate()) return "delegate-pending";
+    // if the user has sent a request to be another wallet's delegate but the other wallet has not yet accepted, then they are a pending delegator
+    if (getHasPendingSentRequestsToBeDelegate()) return "delegator-pending";
+    // if none are true we assume the user has no delegation
+    return "no-delegation";
+  }
+
+  function getHasReceivedRequestsToBeDelegate() {
+    return receivedRequestsToBeDelegate.length > 0;
+  }
+
+  function getHasDelegatorSetEvents() {
+    return delegatorSetEventsForDelegate.length > 0;
+  }
+
+  function getHasPendingReceivedRequestsToBeDelegate() {
+    return getPendingReceivedRequestsToBeDelegate().length > 0;
+  }
+
+  function getHasPendingSentRequestsToBeDelegate() {
+    const pendingSentRequestsToBeDelegate = getPendingSentRequestsToBeDelegate();
+    const mostRecentSentRequestToBeDelegate = pendingSentRequestsToBeDelegate.at(-1);
+    return mostRecentSentRequestToBeDelegate?.delegate !== zeroAddress && pendingSentRequestsToBeDelegate.length > 0;
+  }
+
+  function getPendingReceivedRequestsToBeDelegate() {
+    return (
+      receivedRequestsToBeDelegate
+        ?.filter(
+          (delegateSet) =>
+            !delegatorSetEventsForDelegate?.some(
+              (delegatorSet) =>
+                delegatorSet.delegate == delegateSet.delegate && delegatorSet.delegator == delegatorSet.delegator
+            )
+        )
+        ?.filter(({ delegator }) => !ignoredRequestToBeDelegateAddresses?.includes(delegator)) ?? []
+    );
+  }
+
+  function getPendingSentRequestsToBeDelegate() {
+    return (
```

Comment:
After doing a `setDelegate(0x0)` I'm still having "Pending delegator requests" 

If the "delegator" has set the delegate to `0x0`, ignore the pending delegator requests:

`if(delegate=0x0) return [];`